### ISSUE TITLE
How to write code-base specific lints for correctness and security using Semgrep

### ISSUE THREAD

**Issue by clintgibler** [rails-oceania/roro#274]
## Overall Info

* Length of talk: ~15min.
* Twitter: [@clintgibler](https://twitter.com/clintgibler)
* Headshot: feel free to use from Twitter :)
* Which month: October 2020 or later

## Description / Abstract

Every code base comes with a set of built-in expectations: do things this way, don’t do things that way. These expectations may be related to security (avoid this dangerous function), correctness (if you don’t call foo() before bar(), it’s a bug), performance, robustness, and more.

There are some great tools with out of the box checks you can use, like Brakeman for security and RuboCop for linting and quality.

But what if there are code patterns you want to enforce that are unique to your code bases? Existing tools won’t have these checks, because they’re not general Ruby or framework patterns, they’re specific to you.

You can augment existing tools with custom checks, but this generally requires learning a potentially complicated DSL or API, and becoming familiar with abstract syntax trees (ASTs) and the tool’s architecture. Feasible, but it takes some work.

But what if you could write lints for Ruby that are essentially just Ruby, no complicated upfront learning required?

In this talk, we’ll show how to do just that, using [Semgrep](https://github.com/returntocorp/semgrep), an open source, lightweight static analysis tool. 

We’ll discuss how to:

* Search for Ruby code using patterns that are basically just Ruby (with a few helpful abstractions)
* Find and block security bugs from entering your code
* Scale your team’s productivity: automate the things you always comment in code reviews, and help onboard new engineers with custom checks that teach them calling conventions. 
  * Replace internal docs that people don’t read with checks that scan every PR
* Iteratively explore code bases, finding potential code spots or anomalies; for example, functions whose return value should be checked, or `before_filter`’s that should be consistently applied

You’ll leave this talk with another open source tool in your toolbelt for helping you and your team release higher quality, more secure code, faster and easier.

**Comment by mathewdbutton** [comment]
Sorry for the late reply @clintgibler. This sound fantastic! I can't wait to hear about it, you're locked in for October my friend! Thank you! 🎉

**Comment by clintgibler** [comment]
@mathewdbutton Awesome, very exciting! 🙏 🎉  

Sorry to do this, but would it be possible to do November or December instead, I've had a last minute schedule change and I'm actually going to be on PTO October 13th (the meet-up day, according to [meetup.com](https://www.meetup.com/Ruby-On-Rails-Oceania-Sydney/)).

If Nov or Dec don't work, I could do the following Tues or Wed (Oct 20/21). Let me know, and my apologies!

**Comment by mathewdbutton** [comment]
Yep you've got the right date.

No trouble at all @clintgibler these things happen. We can slot you in for our next meetup on November 10th if that suits better? 

It was all just a stunt to add more anticipation for your talk wasn't it 😄 because it's working.

**Comment by clintgibler** [comment]
Haha! I wish I could claim some sort of marketing genius like that, but unfortunately that's not the case. Just busy 😅  I'm glad to hear the talk sounds interesting though!

November 10th would be perfect, thanks for being accommodating! 🙏

**Comment by mathewdbutton** [comment]
Hey @clintgibler,
I hope you had a nice time off.
I just wanted to check if you are still on for November 10th?

**Comment by clintgibler** [comment]
Hey @mathewdbutton 👋  Thanks, I did :) 

Yes, November 10th still works for me, feel free to create the event / send an invite 🎉 Also, happy to discuss any logistics if that'd be useful: https://twitter.com/clintgibler

**Comment by clintgibler** [comment]
@mathewdbutton Also, just to clarify, would the presentation be 6pm Sydney time? (per the [Meetup page](https://www.meetup.com/Ruby-On-Rails-Oceania-Sydney/))

**Comment by mathewdbutton** [comment]
@clintgibler, excellent question, I really need to give the meetup page a spruce up with more information.

So doors open to the Zoom room at 6pm (AEDT), we do a bit of idle chitchat while we wait for people to turn up and then usually begin at 6:15pm (AEDT), where I or one of the other organisers will do the usual ceremony e.g. who's speaking etc. so speaker presentations will usually start after that at around 6:30pm (AEDT), you're first up so that's when you'll be speaking

**TL;DR**
Doors open at 6pm (AEDT), you'll be speaking at ~6:30pm (AEDT)

Let me know if I've missed something

**Comment by clintgibler** [comment]
@mathewdbutton Sounds great, thanks! Looking forward to it 🎉

**Comment by mathewdbutton** [comment]
Thanks so much for staying up late to present to us @clintgibler that was a really interesting talk.

Would it be okay if I uploaded the recording of your talk to YouTube?

**Comment by clintgibler** [comment]
Np, thanks for having me @mathewdbutton, it was a lot of fun!

And yes, please feel free to upload the video!
